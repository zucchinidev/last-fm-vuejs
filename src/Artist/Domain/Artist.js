export class Artist {
  constructor ({ name, url, associateImages: images } = {}) {
    this.name = name
    this.url = url
    this.images = images
  }
}
