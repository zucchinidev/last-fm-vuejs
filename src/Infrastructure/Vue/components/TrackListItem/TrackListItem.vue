<template>
  <li>
    <h2>
      {{track.album.name}}
    </h2>
    <h3>
      {{track.name}}
    </h3>
    <img :src="imageSrc" alt="track image">
  </li>
</template>
<script>
export default {
  name: 'TrackListItem',
  props: {
    /** @type Track */
    track: {
      type: Object,
      required: true
    }
  },
  computed: {
    imageSrc () {
      let url = ''
      if (Object.keys(this.track).length) {
        const image = this.track.getSmallImage()
        if (image) {
          url = image.url
        }
      }
      return url
    }
  }
}
</script>
<style scoped lang="stylus">
  li
    display block
    margin 10px 0
</style>
