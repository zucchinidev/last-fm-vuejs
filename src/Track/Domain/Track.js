const availableSizes = {
  small: 64,
  medium: 300,
  big: 640
}

export class Track {
  constructor ({ album, duration, name, popularity, images, id, previewUrl } = {}) {
    this.id = id
    this.album = album
    this.duration = duration
    this.name = name
    this.popularity = popularity
    this.images = images
    this.previewUrl = previewUrl
  }

  getSmallImage () {
    return this.images.find(image => image.width === availableSizes.small)
  }

  getBigImage () {
    return this.images.find(image => image.width === availableSizes.big)
  }

  getPreviewUrl () {
    return this.previewUrl
  }
}
