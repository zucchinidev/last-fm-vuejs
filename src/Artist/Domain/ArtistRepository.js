import { HttpRepository } from '../../Infrastructure/http/HttpRepository'
import config from '../../Infrastructure/lastFmApi/config'
import { Artist } from './Artist'
const urls = {
  getArtistsByCountry: ''
}

const configuration = {
  slug: 'contacts',
  baseURL: 'http://ws.audioscrobbler.com/2.0/'
}
export const ArtistRepository = {
  async getTopArtistsByCountry (country = 'spain') {
    const params = {
      method: 'geo.gettopartists',
      country,
      api_key: config.apiKey,
      format: 'json'
    }
    const { data } = await HttpRepository.get(urls.getArtistsByCountry, params, configuration)
    return data.topartists.artist.map(artist => new Artist({
      images: artist.image.map(image => ({url: image['#text'], size: image.size})),
      url: artist.url,
      name: artist.name
    }))
  }
}
