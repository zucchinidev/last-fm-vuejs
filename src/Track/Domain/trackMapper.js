import { Track } from './Track'

export default (data) => {
  const artists = (data.album.artists || []).map(artist => artist.name)
  const images = data.album.images || {}
  return new Track({
    album: {
      artists,
      name: data.album.name,
      release: data.album.release_date
    },
    images,
    duration: data.duration_ms,
    name: data.name,
    popularity: data.popularity
  })
}
