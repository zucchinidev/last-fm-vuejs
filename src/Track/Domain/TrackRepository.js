import { HttpRepository } from '../../Infrastructure/http/HttpRepository'
import trackMapper from './trackMapper'

const urls = {
  search: 'search'
}

const configuration = {
  baseURL: process.env.API_MUSIC_BASE_URL
}
export const TrackRepository = {
  async searchTrack (searchTerm) {
    const params = {
      q: searchTerm,
      type: 'track'
    }
    const { data } = await HttpRepository.get(urls.search, params, configuration)
    return data.tracks.items.map(trackMapper)
  }
}
