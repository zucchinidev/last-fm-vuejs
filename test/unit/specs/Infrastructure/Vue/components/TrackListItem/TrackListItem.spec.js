import TrackListItem from '@Vue/components/TrackListItem/TrackListItem'
import { WrapComponent } from '../../../../../helpers'
import { Track } from '../../../../../../../src/Track/Domain/Track'
import * as fixture from '../../../../../fixtures/trackFixture'
import TrackListItemPageObject from './TrackListItemPageObject'

const { track: trackFixture } = fixture
describe('TrackListItem.vue', () => {
  let wrapper
  /** @type Track */
  let track
  /** @type TrackListItemPageObject */
  let page
  beforeEach(() => {
    track = new Track(trackFixture)
    wrapper = WrapComponent(TrackListItem).withProps({track}).mount()
    page = new TrackListItemPageObject(wrapper)
  })
  it('should render correct contents', () => {
    expect(page.getAlbumTitle()).toEqual(trackFixture.album.name)
    expect(page.getTrackTitle()).toEqual(trackFixture.name)
    expect(page.getPoster().src).toEqual(track.getSmallImage().url)
    expect(page.getPoster().alt).toEqual('track image')
  })

  it('must receive a track as property', () => {
    expect(wrapper.props().track).toEqual(track)
  })

  it('should has correct html', function () {
    page.matchSnapshot()
  })
})
