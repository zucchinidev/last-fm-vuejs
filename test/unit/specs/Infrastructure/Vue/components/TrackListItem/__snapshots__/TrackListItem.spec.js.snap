// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackListItem.vue should has correct html 1`] = `
<li>
    <h2>
        Esa Muchacha
    </h2>
    <h3>
        Esa Muchacha
    </h3>
    <img src="https://i.scdn.co/image/31dd0b4e46c5fb0feb1e02c2f98d07e76ca5ab6c" alt="track image">
</li>
`;
